### ISSUE TITLE
All first lines uncovered when using `c8` with `tsx --test`

### ISSUE THREAD

**Issue by rozzilla** [privatenumber/tsx#564]
### Acknowledgements

- [X] I read the documentation and searched existing issues to avoid duplicates
- [X] I am reporting a confirmed bug (not seek debugging help)
- [X] I understand this is a collaborative project and relies on my contributions
- [X] I read and understood the [Contribution guide](https://github.com/privatenumber/tsx/blob/develop/CONTRIBUTING.md)

### Minimal reproduction URL

https://github.com/rozzilla/c8-tsx-coverage-issue

### Problem & expected behavior (under 200 words)

As you can see from the repro, even though I have a test covering the whole file, the coverage result return the 1st line as uncovered

### Bugs are expected to be fixed by those affected by it

- [ ] I'm interested in working on this issue

### Compensating engineering work financially will speed up resolution

- [ ] I'm willing to offer financial support

**Comment by privatenumber** [comment]
tsx is just using the Node APIs so I'm not sure if this is a bug in tsx or if there's even anything we can do on our end.

This may be fixed by:
https://github.com/nodejs/node/pull/53315

**Comment by rozzilla** [comment]
To better explain the issue, this doesn't happen with `nyc`

Il Gio 6 Giu 2024, 08:49 Hiroki Osame ***@***.***> ha scritto:

**Comment by privatenumber** [comment]
Not sure if I understand how that better explains the issue...

If you want to get to the bottom of this, I recommend inlining the sourcemap and filing an issue with Node.

**Comment by rozzilla** [comment]
I also found another issue.
If I [run the coverage](https://github.com/rozzilla/c8-tsx-coverage-issue/blob/main/package.json#L7) with `nyc` on the latest `tsx` version (so `4.15.6`) I get no coverage. If I downgrade `tsx` to `4.7.2` I get the coverage. 🤔 Do you know @privatenumber what changed meanwhile that is causing this regression?

**Comment by privatenumber** [comment]
Locking as this seems blocked by Node. Feel free to open an issue there if it doesn't exist yet.

**Comment by privatenumber** [comment]
This is a `v8-to-istanbul` sourcemap-remapping bug, not uncovered user code, and there is no appropriate fix in tsx.

tsx transforms this file as CommonJS, so esbuild generates export helpers. V8 records uncovered branches inside that generated code, which has no corresponding TypeScript mapping.

`v8-to-istanbul` cannot find a mapping for one helper range, so its [least-upper-bound fallback](https://github.com/istanbuljs/v8-to-istanbul/blob/v9.3.0/lib/source.js#L205-L217) selects the first mapping in the file. That incorrectly attributes the generated branch to `index.ts:1`.

This still reproduces with current tsx, Node 24.15.0, and c8 12.0.0. The raw V8 data confirms that the original `sum` function is fully executed:

- Default c8 reports 77.77% branch coverage and identifies line 1.
- ESM output, which does not need the CommonJS bridge, reports 100%.
- c8's Monocart reporter uses the same raw V8 data and reports 100%.

`v8-to-istanbul` has not shipped a functional update since June 2024. Its proposed sourcemap improvements in istanbuljs/v8-to-istanbul#244 were closed in favor of a different converter.

Given the upstream maintenance state and the working Monocart alternative, I'm closing this as an upstream coverage-conversion limitation.
